sdks/java/io/tika/src/main/java/org/apache/beam/sdk/io/tika/TikaSource.java

```diff
@@ -0,0 +1,466 @@
+      if (metadataIterator == null) {
+        // Check if some content is already available
+        current = tikaHandler.getCurrent();
+
+        if (current == null && !Boolean.TRUE.equals(spec.getParseSynchronously())) {
+          long maxPollTime = 0;
+          long configuredMaxPollTime = spec.getQueueMaxPollTime() == null
+              ? TikaIO.Read.DEFAULT_QUEUE_MAX_POLL_TIME : spec.getQueueMaxPollTime();
+          long configuredPollTime = spec.getQueuePollTime() == null
+              ? TikaIO.Read.DEFAULT_QUEUE_POLL_TIME : spec.getQueuePollTime();
+
+          // Poll the queue till the next piece of data is available
+          while (current == null && maxPollTime < configuredMaxPollTime) {
+            boolean docEnded = tikaHandler.waitForNext(configuredPollTime);
+            current = tikaHandler.getCurrent();
+            // End of Document ?
+            if (docEnded) {
+              break;
+            }
+            maxPollTime += spec.getQueuePollTime();
+          }
+        }
+        // No more content ?
+        if (current == null && Boolean.TRUE.equals(spec.getReadOutputMetadata())) {
+          // Time to report the metadata
+          metadataIterator = Arrays.asList(tikaMetadata.names()).iterator();
+        }
+      }
+
+      if (metadataIterator != null && metadataIterator.hasNext()) {
+          String key = metadataIterator.next();
+          // The metadata name/value separator can be configured if needed
+          current = key + "=" + tikaMetadata.get(key);
+      }
+      return current != null;
+    }
+
+    @Override
+    public String getCurrent() throws NoSuchElementException {
+      if (current == null) {
+        throw new NoSuchElementException();
+      }
+      return current;
+    }
+
+    @Override
+    public void close() throws IOException {
+      if (execService != null) {
+          execService.shutdown();
+      }
+    }
+
+    ExecutorService getExecutorService() {
+      return execService;
+    }
+
+    @Override
+    public BoundedSource<String> getCurrentSource() {
+      return source;
+    }
+  }
+
+  /**
+   * Tika Parser Content Handler.
+   */
+  static class ContentHandlerImpl extends DefaultHandler {
+    private Queue<String> queue = new ConcurrentLinkedQueue<>();
+    private volatile boolean documentEnded;
+    private volatile Exception parseException;
+    private volatile String current;
+    private int minTextLength;
+
+    @Override
+    public void characters(char ch[], int start, int length) throws SAXException {
+      String value = new String(ch, start, length).trim();
+      if (!value.isEmpty()) {
+        if (minTextLength <= 0) {
+          queue.add(value);
+        } else {
+          current = current == null ? value : current + " " + value;
+          if (current.length() >= minTextLength) {
+            queue.add(current);
+            current = null;
+          }
+        }
+      }
+    }
+
+    public void setParseException(Exception ex) {
```

Comment:
Does this happen in the same thread or in a different thread than reading this variable?